File: windows/src/ConnectorControl.Core/State/EditTarget.cs
namespace ConnectorControl.Core.State;

/// <summary>
/// What an editor window edits. An existing connector's id is its collection, <c>\u001F</c>, then
/// its name, so there is one window per connector per collection; a new one's is a fresh GUID.
///
/// Mirror: Sources/ConnectorControlState/EditTarget.swift
/// </summary>
/// <param name="Collection">
/// Which collection this window edits. Always named: a window that followed whichever collection
/// is active would move its read-only state, its header and its save target whenever the active
/// collection changed under it. It sits before <paramref name="ForcesRemote"/>, unlike the
/// Mac's initializer, because C# puts a required parameter ahead of an optional one.
/// </param>
public sealed record EditTarget(string Id, string Name, McpEntry Entry, bool IsNew, string Collection, bool ForcesRemote = false)
{
    public static EditTarget Existing(string name, McpEntry entry, string collection) =>
        new(Identifier(name, collection), name, entry, IsNew: false, Collection: collection);

    /// <summary>Add-Remote flow: the template has an empty URL that Detect() can't classify, so the remote form is forced explicitly.</summary>
    public static EditTarget NewRemote(RemoteLaunchStyle style, string collection) =>
        new(Guid.NewGuid().ToString(), "", new McpEntry(RemotePattern.Make("", style)), IsNew: true, ForcesRemote: true,
            Collection: collection);

    /// <summary>
    /// The same connector in two collections is two windows, so the collection is part of the
    /// identity. The separator is a control character rather than a slash or a dot, neither of
    /// which a collection or a connector name is stopped from containing.
    /// </summary>
    internal static string Identifier(string name, string collection) => collection + '\u001F' + name;

    public const string AddTitle = "Add Connector";

    public static string EditTitle(string name) => $"Edit “{name}”";

    /// <summary>Fixed at open time.</summary>
    public string WindowTitle => IsNew ? AddTitle : EditTitle(Name);

    /// <summary>
    /// One window per connector per collection: two targets are the same window when their ids
    /// match, whatever the entry held when each was made. WindowRegistry keys its editors by the
    /// id, so a connector switched on or off since its editor opened still brings that editor
    /// forward rather than opening a second. The Mac's editor window group finds its window by
    /// this same equality.
    /// </summary>
    public bool Equals(EditTarget? other) => other is not null && string.Equals(Id, other.Id, StringComparison.Ordinal);

    public override int GetHashCode() => StringComparer.Ordinal.GetHashCode(Id);
}

File: Sources/ConnectorControlState/EditTarget.swift
import Foundation
import ConnectorControlCore

/// What an editor window edits. An existing connector's id is its collection, `\u{001F}`, then
/// its name, so there is one window per connector per collection; a new one's is a fresh UUID.
/// Codable and Hashable for the SwiftUI WindowGroup value.
///
/// Mirror: windows/src/ConnectorControl.Core/State/EditTarget.cs
public struct EditTarget: Identifiable, Codable, Hashable, Sendable {
    /// The editor `WindowGroup`'s id — shared by `ConnectorControlApp`'s declaration and
    /// `CollectionsWindowView`'s `openWindow` calls so they cannot drift apart.
    public static let editorWindowID = "editor"
    public static let addTitle = "Add Connector"

    public static func editTitle(_ name: String) -> String { "Edit “\(name)”" }

    public let id: String
    public var name: String
    public var entry: MCPEntry
    public var isNew: Bool
    public var forcesRemote: Bool
    /// Which collection this window edits. Always named: a window that followed whichever
    /// collection is active would move its read-only state, its header and its save target
    /// whenever the active collection changed under it.
    public var collection: String

    public init(id: String, name: String, entry: MCPEntry, isNew: Bool, forcesRemote: Bool = false,
                collection: String) {
        self.id = id
        self.name = name
        self.entry = entry
        self.isNew = isNew
        self.forcesRemote = forcesRemote
        self.collection = collection
    }

    public static func existing(name: String, entry: MCPEntry, in collection: String) -> EditTarget {
        EditTarget(id: identifier(name, in: collection), name: name, entry: entry, isNew: false,
                   collection: collection)
    }

    /// Add-Remote flow: the template has an empty URL that detect() cannot
    /// classify, so the remote form is forced explicitly.
    public static func newRemote(in collection: String) -> EditTarget {
        EditTarget(id: UUID().uuidString, name: "",
                   entry: MCPEntry(config: RemotePattern.make(url: "")), isNew: true, forcesRemote: true,
                   collection: collection)
    }

    /// The same connector in two collections is two windows, so the collection is part of the
    /// identity. The separator is a control character rather than a slash or a dot, neither of
    /// which a collection or a connector name is stopped from containing.
    static func identifier(_ name: String, in collection: String) -> String {
        collection + "\u{001F}" + name
    }

    /// Fixed at open time.
    public var windowTitle: String { isNew ? EditTarget.addTitle : EditTarget.editTitle(name) }
}
